style(application): adds tsfmt and tslint to the build

priley86 opened the pull request:
* adds [tsfmt](https://github.com/vvakame/typescript-formatter) to format all Typescript files in accordance with the current tsconfig/tslint
* adds tslint and tslint fix. There are still some lint errors... do you want me to fix the following? 

```
src/app/card/card-filter/card-filter.component.ts[26, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/card/card-filter/card-filter.component.ts[31, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.base.ts[45, 13]: comment must start with a space
src/app/chart/chart.defaults.ts[8, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[9, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[10, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[11, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[12, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[13, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[14, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[15, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[16, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[17, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[18, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[19, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[20, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[21, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/chart.defaults.ts[22, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/sparkline/examples/sparkline-example.component.ts[14, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/sparkline/examples/sparkline-example.component.ts[18, 3]: Declaration of public instance field not allowed after declaration of private instance field. Instead, this should come at the beginning of the class/interface.
src/app/chart/sparkline/sparkline.component.spec.ts[36, 1]: Exceeds maximum line length of 120
src/app/chart/sparkline/sparkline.component.ts[130, 20]: Missing radix parameter
src/app/chart/sparkline/sparkline.component.ts[131, 17]: Missing radix parameter
src/app/chart/sparkline/sparkline.component.ts[77, 1]: Exceeds maximum line length of 120
src/app/chart/sparkline/sparkline.component.ts[99, 1]: Exceeds maximum line length of 120
src/app/chart/sparkline/sparkline.component.ts[133, 1]: Exceeds maximum line length of 120
src/app/chart/sparkline/sparkline.component.ts[174, 26]: comment must start with a space
src/app/chart/sparkline/sparkline.component.ts[182, 26]: comment must start with a space
src/app/navigation/vertical-navigation.component.spec.ts[377, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.spec.ts[380, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.spec.ts[383, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.spec.ts[397, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.spec.ts[400, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.spec.ts[417, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.spec.ts[420, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.spec.ts[423, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.spec.ts[433, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.spec.ts[436, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.spec.ts[456, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.spec.ts[459, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.spec.ts[462, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.ts[456, 3]: Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
src/app/navigation/vertical-navigation.component.ts[475, 3]: Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
src/app/navigation/vertical-navigation.component.ts[506, 3]: Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
src/app/navigation/vertical-navigation.component.ts[513, 3]: Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
src/app/navigation/vertical-navigation.component.ts[535, 3]: Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
src/app/navigation/vertical-navigation.component.ts[553, 3]: Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
src/app/navigation/vertical-navigation.component.ts[572, 3]: Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
src/app/navigation/vertical-navigation.component.ts[584, 3]: Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
src/app/navigation/vertical-navigation.component.ts[605, 3]: Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
src/app/navigation/vertical-navigation.component.ts[626, 3]: Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
src/app/navigation/vertical-navigation.component.ts[647, 3]: Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
src/app/navigation/vertical-navigation.component.ts[668, 3]: Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
src/app/navigation/vertical-navigation.component.ts[687, 3]: Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
src/app/navigation/vertical-navigation.component.ts[161, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.ts[572, 1]: Exceeds maximum line length of 120
src/app/navigation/vertical-navigation.component.ts[248, 11]: comment must start with a space
src/app/navigation/vertical-navigation.component.ts[278, 7]: comment must start with a space
src/app/navigation/vertical-navigation.component.ts[287, 7]: comment must start with a space
src/app/sample/sample.component.spec.ts[26, 11]: comment must start with a space
src/app/sample/sample.component.spec.ts[28, 11]: comment must start with a space
src/demo/app-routing.module.ts[9, 1]: Exceeds maximum line length of 120
src/demo/app-routing.module.ts[10, 1]: Exceeds maximum line length of 120
src/demo/app.module.ts[20, 1]: Exceeds maximum line length of 120
```

priley86 committed:
style(application): adds tsfmt and tslint to the build and fixes formatting

dlabrecq commented on src/app/empty-state/empty-state.component.spec.ts:
```diff
@@ -39,8 +39,8 @@ describe('Empty state component - ', () => {
-        'impression that helps users to achieve their goals. It should be used when a view is empty because no ' +
-        'objects exists and you want to guide the user to perform specific actions.',
+      'impression that helps users to achieve their goals. It should be used when a view is empty because no ' +
+      'objects exists and you want to guide the user to perform specific actions.',
```
Odd that you cannot have a line-wrapping indentation here?

dlabrecq commented on src/app/chart/sparkline/examples/sparkline-example.module.ts:
```diff
@@ -8,16 +8,16 @@ import { SparklineExampleComponent } from './sparkline-example.component';
 })
 export class SparklineExampleModule {
-  constructor() {}
+  constructor() { }
```
Which rule states there must be a space here?

priley86 replied:
Not sure - will look to find it...

dlabrecq commented on src/app/filter/filter-results.component.ts:
```diff
@@ -92,7 +92,7 @@ export class FilterResultsComponent implements DoCheck, OnInit {
       if (appliedFilter.field.title !== filter.field.title
-          || appliedFilter.value !== filter.value) {
+        || appliedFilter.value !== filter.value) {
         newFilters.push(appliedFilter);
```
Don't like how the indentation was removed here. I've always indented this way (a Java standard) for readability.

dlabrecq commented on src/app/list/basic-list/examples/list-basic-example.component.ts:
```diff
@@ -172,7 +172,7 @@ export class ListBasicExampleComponent implements OnInit {
    */
   getActionConfig(item: any, actionButtonTemplate: TemplateRef<any>,
-      startButtonTemplate: TemplateRef<any>): ActionConfig {
+    startButtonTemplate: TemplateRef<any>): ActionConfig {
```
Same here. The line wrapping indentation is removed.

joshuawilson replied:
This one is a bit more complicated for me. I think it should either be a normal 2 space indent or it is indented to be inline with the same context above. That would be about 17-18 space indent. I have done both but not an arbitrary 4 space indent. But that is just my experience.

dlabrecq replied:
It's a Java style convention for wrapping lines that I've always liked and followed. It's my preference, but can live without it.

dlabrecq commented on src/app/sort/sort.component.ts:
```diff
@@ -89,7 +89,7 @@ export class SortComponent implements DoCheck, OnInit {
     let iconStyleClass: string;
     if (this.currentField && this.currentField.sortType
-        && this.currentField.sortType === 'numeric') {
+      && this.currentField.sortType === 'numeric') {
```
Another one...

dlabrecq reviewed (changes requested):
In general, I don't mind committing these fixes. However, I'm not a fan of the build changing the code automatically. For example, I have some line wrapping indentation that was removed. Wondering if ts:lint-fix should be run manually?

Noticed that the Travis build also fails unless everything is fixed, including the long lines. Unfortunately, the release will fail until that is addressed as well.

priley86 commented:
@dlabrecq agree that the build does not need to format the code for us - only validate that the formatter has been run (via the tslint validation rules). This PR introduces that (i.e.: I only added `tslint:format` and `ts:lint` to the build step (it might confuse that `tslint:format` is only validating and that `tslint:format:fix` is fixing, so I'd be perfectly fine changing the names of the scripts).

dlabrecq commented:
Thanks, @priley86. I have a couple more comments after running the changes locally.

1. Perhaps ts:format would be better as ts:verify? It appears the npm command is running '--verify'.

2. I noticed some odd behavior during testing. For example, I added an extra space, but it didn't tell me what the problem was or even a line number. The build just failed with the message below.

"/Users/dlabrecq/Work/patternfly/repos/patternfly-ng/src/app/filter/filter-results.component.ts is not formatted"

Is there a way we can get a better description? I'd like to see the error first, perhaps even fix the problem myself, before running any formatters.

3. The linter is also not run for the start:demo target, which is what I mostly use during development. Wondering if we will want to add this there as well? 

4. I also noticed npm-debug.log and package-lock.json files which probably should be removed. We don't have a shrinkwrap, so I'm not sure how package-lock.json got there?

priley86 commented:
@dlabrecq so did a bit more research this morning here.

1. updated to `ts:format:verify` and `ts:format:fix`. These are just related to formatting so I figured that would be appropriate. Not really attached to this at all though.
2. I tried using the `--verbose` flag with tsfmt. It gives output on what configuration was used (such as below) to verify the file, but it does not give the line number the error occurred. I agree this is annoying and would propose that we open an issue upstream about this. Should I open an issue?
```
editorconfig: 
file: /Users/priley/GitHub/patternfly-ng/src/app/filter/filter-results.component.ts
{
  "charset": "utf-8",
  "indent_style": "space",
  "indent_size": 2,
  "end_of_line": "lf",
  "insert_final_newline": true,
  "trim_trailing_whitespace": true,
  "tab_width": 2
}
```
It looks like the linter (tslint) will still report such errors though so this may be a non-issue... as the goal is to make the formatter to integrate with the linter.

3. It would be ideal if the formatter could integrate with the editor so you get real time feedback/real time formatting on every save (this is how Prettier is configured for comparison). I don't know that you'd necessarily need to run an npm script everytime a file changes (it may be slow formatting all files that way and I don't see a way to specify a file in a livereload like way). It does appear though that some IDEs are making an effort to support typescript-formatter as a built in plugin. I haven't tried this yet with VSCode or Webstorm, but it appears VSCode is making an effort to support this typescript-formatter. For that reason, I'd probably say tsmt is a better option long term for TS. Link: https://github.com/Microsoft/vscode/issues/17627
4. I've removed package-lock.json. Was not aware you are not using NPM5 yet here (or maybe you have it defined some place else downstream?).

dlabrecq commented:
Thank you for the update. Will need to add ts:format:verify to the build target as ts:format no longer exists.

My only hang up is the lack of error details. The CSS linter; for example, will generate errors like this:

src/app/list/basic-list/list.component.less
 9:21  :heavy_multiplication_x:  Expected newline after "{" of a multi-line block 
block-opening-brace-newline-after

For kicks, I removed ts:format:verify and just used tslint. The error messages below are what I'm looking for.

src/app/action/action.component.ts[37, 21]: expected space after colon in property-declaration
src/app/action/action.component.ts[37, 21]: missing whitespace

That said, I'm thinking we can just remove tsfmt, but have a couple questions first.

1. Do we really want both typescript-formatter and tslint -- seems one or the other would do?
2. What is ts:format:verify doing for us that tslint is not (e.g., it appears tsfmt is just using the same rules from tslint.json)?
3. Do we need ts:format:verify if we don't intend to reformat code?
4. Could we not run ts:format:verify after ts:lint?

priley86 commented:
@dlabrecq I've left `ts:format:verify` in the build step for now. I think this PR can wait for a while until we are comfortable with a typescript formatter (not sure if this is the one or not, but seems to have the most community support a.t.m.). Are there any upstream issues we'd like to open to make the formatter more reliable?  Please proceed with tslint in a separate PR if you like...

1. The formatter ensures consistently formatted code by automating the formatting. The linter checks these formats pass its own rules. This formatter tries to integrate with those rules. The goal of the formatter is to automate formatting and reduce time we spend manually doing this ourselves. This formatter also integrates with some IDEs it appears, and can also be run via the script...
2. ts:format:verify just ensures the formatter has been run by the developer with the same configurations used in the ts:format:fix script. I agree that build step does not need to actually format the code, just verify that it has been formatted...
3. If we don't intend to format the code, then this PR can wait...
4. Yes. I think it could run after the lint rule checks. 

Will await any more thoughts the team has on this.

dlabrecq commented:
@priley86 @dgutride Anyone mind if I close this? We have a TS linter in place now, just without the formatter.
